### PR TITLE
Create a copy

### PR DESCRIPTION
**Pull request by rgarner**
# [Create a New Form and Info Sheet from a Saved Form and Info Sheet (aka Research Session)](https://trello.com/c/qYBbtDC9/87-create-a-new-form-and-info-sheet-from-a-saved-form-and-info-sheet-aka-research-session)

> When I have a second (related) research session,
> I want to create a new consent form ,
> so I can avoid typing everything from scratch again.

## Screenshots

screen shot 2017-11-07 at 11 54 12

screen shot 2017-11-07 at 11 55 20

## Things to consider that aren't acceptance criteria

* The visibility of share/new buttons during the copy interstitial
* Should we have Stay/Leave behaviour for leaving edited steps when using New/Create a copy?

### CODE REVIEW

File: features/step_definitions/aids_for_devops_steps.rb [hunk 1]
```diff
@@ -27,5 +27,6 @@
 end
 
 When(/^I start to create a form$/) do
+  fill_in 'Please name your research session', with: 'Any old pony'
```
Comment (joelcarr) [inline features/step_definitions/aids_for_devops_steps.rb]: We should use a realistic example for context for the next maintainers

Comment (rgarner) [reply features/step_definitions/aids_for_devops_steps.rb]: It means we don't care about the value in this context, just that one exists – perhaps that could be more explicit.

**Comment by rgarner** [comment]
Currently trying to rebase this but the presenter upgrade has caused a kerfuffle

File: app/controllers/research_sessions_controller.rb [hunk 2]
```diff
@@ -12,6 +21,32 @@ def create
...
+
+  # Only ever called in the context of copying a new form
+  def update
+    @research_session = current_research_session.tap do |session|
+      session.name = session_creation_params[:name]
+      session.set_slug_from_name!
+      if session.save
+        redirect_to(research_session_preview_path(session))
+      else
+        render 'edit'
```
Comment (rgarner) [inline app/controllers/research_sessions_controller.rb]: Oops. No edit any more

File: app/views/research_sessions/new.html.erb [hunk 3]
```diff
@@ -9,9 +9,10 @@
   <%= render 'shared/error_summary' %>
 
   <%= barnardos_form_with model: @research_session do |form| %>
-    <%= form.labelled_text_field :name, text_options: { autofocus: true } %>
-    <p class="button-bar">
-      <%= form.submit 'Create new form', class: 'button' %>
-    </p>
+      <%= form.hidden_field :id %>
```
Comment (rgarner) [inline app/views/research_sessions/new.html.erb]: Still required

File: config/routes.rb [hunk 4]
```diff
@@ -2,9 +2,10 @@
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root to: 'research_sessions#new'
 
-  resources :research_sessions, only: [:new, :create], path: 'research-sessions' do
+  resources :research_sessions, only: [:new, :create, :update], path: 'research-sessions' do
```
Comment (rgarner) [inline config/routes.rb]: Update still required?

File: app/views/research_sessions/new.html.erb [hunk 5]
```diff
@@ -9,9 +9,9 @@
...
 
   <%= barnardos_form_with model: @research_session do |form| %>
-    <%= form.labelled_text_field :name, text_options: { autofocus: true } %>
-    <p class="button-bar">
-      <%= form.submit 'Create new form', class: 'button' %>
-    </p>
+      <%= form.labelled_text_field :name, text_options: { autofocus: true } %>
+      <p class="button-bar">
+        <%= form.submit 'Create new form', class: 'button' %>
+      </p>
```
Comment (joelcarr) [inline app/views/research_sessions/new.html.erb]: `    `

**Pull request merged**